pkg/discovery/type/git/git.go

```diff
@@ -0,0 +1,138 @@
+Copyright © 2022 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"strings"
+
+	git "github.com/go-git/go-git/v5"
+	provv1 "github.com/rancher-sandbox/rancheros-operator/pkg/apis/rancheros.cattle.io/v1"
+)
+
+type githubOptions struct {
+	repository string
+	subdir     string
+	ctx        context.Context
+}
+
+type githubSetting func(g *githubOptions) error
+
+// WithRepository sets a git repository to scan releases against
+func WithRepository(s string) githubSetting { //nolint:golint,revive
+	return func(g *githubOptions) error {
+		g.repository = s
+		return nil
+	}
+}
+
+// WithSubpath sets a repository
+func WithSubpath(s string) githubSetting { //nolint:golint,revive
+	return func(g *githubOptions) error {
+		g.subdir = s
+		return nil
+	}
+}
+
+// WithContext sets a context for the discovery action
+func WithContext(ctx context.Context) githubSetting { //nolint:golint,revive
+	return func(g *githubOptions) error {
+		g.ctx = ctx
+		return nil
+	}
+}
+
+func (g *githubOptions) apply(opts ...githubSetting) error {
+	for _, o := range opts {
+		if err := o(g); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewReleaseFinder returns a new Github release finder discovery with the required settings
+func NewReleaseFinder(opts ...githubSetting) (*releaseFinder, error) { //nolint:golint,revive
+	o := &githubOptions{
+		ctx: context.Background(),
+	}
+
+	err := o.apply(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &releaseFinder{
+		opts: *o,
+	}, nil
+}
+
+type releaseFinder struct {
+	opts githubOptions
+}
+
+// Discovery retrieves ManagedOSVersion from github releases
+func (f *releaseFinder) Discovery() (res []*provv1.ManagedOSVersion, err error) {
+
+	opts := &git.CloneOptions{
+		URL: f.opts.repository,
+	}
+
+	temp, err := os.MkdirTemp("", "rf")
+	if err != nil {
+		return res, err
+	}
+
+	defer os.RemoveAll(temp)
+
+	_, err = git.PlainClone(temp, false, opts)
+	if err != nil {
+		return res, fmt.Errorf("error while failed cloning repo: %w", err)
+	}
+
+	err = filepath.Walk(filepath.Join(temp, f.opts.subdir),
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !strings.HasSuffix(path, "json") {
+				return nil
+			}
+			v := &provv1.ManagedOSVersion{}
+			dat, err := ioutil.ReadFile(path)
+			if err != nil {
+				return err
+			}
+			err = json.Unmarshal(dat, v)
```

Comment:
missing some logging here to know that it failed on unmarshalling? Or is this on purpose in case there is other json files in the repo that we fail to unmarshall? Im just thinking that it would be useful to have an error in case the release file is wrongly formatted and we cannot get it

Reply:
so yup here It was on purpose so you could have other files in the repository (I'm thinking at github actions, eventual scripts, some README, etc),  in that way acts as best-effort. 

Maybe we can collect logs here, but if we don't fail explicitly it's not going to surface to the user quite easily. I have no strong opinion on both, maybe we should just have a common suffix (e.g. `.version.json` ) so we differentiate from all the other files in the repo, wdyt?

Reply:
I have no idea. I mean, Im guessing that normally you would have your version files in a subfolder with those in it to be a bit more organized, no one wants to see 30 json files lying around with other files on it, so this problem should not arise.